ipaserver/plugins/hbacrule.py

```diff
@@ -425,67 +425,61 @@ def execute(self, cn, **options):
-        ),
-    )
+    member_attributes = ['membertimerule']
+    member_count_out = ('%i object added.', '%i objects added.')
 
-    def execute(self, cn, **options):
+    def execute(self, *args, **options):
         ldap = self.obj.backend
+        dn = self.obj.get_dn(*args, **options)
+        assert(isinstance(dn, DN))
 
-        dn = self.obj.get_dn(cn)
-
-        entry_attrs = ldap.get_entry(dn, ['accesstime'])
-        entry_attrs.setdefault('accesstime', []).append(
-            options['accesstime']
-        )
         try:
-            ldap.update_entry(entry_attrs)
-        except errors.EmptyModlist:
-            pass
+            entry_attrs = ldap.get_entry(dn, ['objectclass'])
         except errors.NotFound:
-            self.obj.handle_not_found(cn)
+            self.obj.handle_not_found(*args)
+        objclass_updated = False
+        if ('ipahbacrulev2' not in
+                (o.lower() for o in entry_attrs['objectclass'])):
+            entry_attrs['objectclass'] = ['ipaassociation', 'ipahbacrulev2']
+            ldap.update_entry(entry_attrs)
+            objclass_updated = True
 
-        return dict(result=True)
+        try:
+            result = super(hbacrule_add_timerule, self).execute(*args, **options)
+        except Exception as e:
+            self.log.error("Failed to add a timerule: {err}".format(err=e))
+            if objclass_updated:
+                entry_attrs['objectclass'] = ['ipaassociation', 'ipahbacrule']
+                ldap.update_entry(entry_attrs)
+            raise
+        return result
 
 
-# @register()
-class hbacrule_remove_accesstime(LDAPQuery):
-    """
-    Remove access time to HBAC rule.
-    """
-    takes_options = (
-        AccessTime('accesstime?',
-            cli_name='time',
-            label=_('Access time'),
-        ),
-    )
+@register()
+class hbacrule_remove_timerule(LDAPRemoveMember):
+    __doc__ = _('Remove users and groups from an HBAC rule.')
 
-    def execute(self, cn, **options):
-        ldap = self.obj.backend
+    member_attributes = ['membertimerule']
+    member_count_out = ('%i object removed.', '%i objects removed.')
 
-        dn = self.obj.get_dn(cn)
+    def execute(self, *args, **options):
+        result = super(hbacrule_remove_timerule, self).execute(*args,
+                                                               **options)
+        dn = result['result']['dn']
+        assert(isinstance(dn, DN))
+        timerules = result['result'].get('membertimerule_timerule', [])
 
-        entry_attrs = ldap.get_entry(dn, ['accesstime'])
-        try:
-            entry_attrs.setdefault('accesstime', []).remove(
-                options['accesstime']
-            )
+        ldap = self.obj.backend
+        entry_attrs = ldap.get_entry(dn, ['objectclass'])
+        if (len(timerules) == 0 and 'ipahbacrulev2' in
+               (o.lower() for o in entry_attrs['objectclass'])):
+            entry_attrs['objectclass'] = ['ipaassociation', 'ipahbacrule']
```

Comment:
1. again touch only 'ipahbacrule' and 'ipahbacrulev2' objectclasses, you should not care about others

2. be more pythonic please
```
if not timerules and ...
```